rapidpro/models/routers.py

```diff
@@ -0,0 +1,174 @@
+import logging
+import string
+from random import random
+
+from rapidpro.models.common import Exit
+from rapidpro.utils import generate_new_uuid
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRouter:
+    def __init__(self, operand, result_name=None, wait_for_message=False):
+        self.type = None
+        self.categories = []
+        self.default_category_uuid = None
+        self.cases = []
+        self.operand = operand
+        self.wait_for_message = wait_for_message
+        self.result_name = result_name
+
+    def set_result_name(self, result_name):
+        self.result_name = result_name
+
+    def _get_category_or_none(self, category_name):
+        result = [c for c in self.categories if c.name == category_name]
+        if result:
+            return result[0]
+
+    def _has_default_category(self):
+        result = [c for c in self.categories if c.is_default]
+        return bool(result)
+
+    def _add_category(self, category_name, destination_uuid, is_default):
+        if self._has_default_category() and is_default:
+            logger.warning(f'Overwriting default category for Router {self.uuid}')
+
+        category = RouterCategory(category_name, destination_uuid, is_default)
+
+        if is_default:
+            self.default_category_uuid = category.uuid
+
+        self.categories.append(category)
+        return self.categories[-1]
+
+    def _get_case_or_none(self, comparison_type, arguments, category_uuid):
+        for case in self.cases:
+            if case.comparison_type == comparison_type \
+                    and case.arguments == arguments \
+                    and case.category_uuid == category_uuid:
+                return case
+
+    def _add_case(self, comparison_type, arguments, category_uuid):
+        case = RouterCase(comparison_type, arguments, category_uuid)
+        self.cases.add(case)
+        return self.cases[-1]
+
+    def get_or_create_case(self, comparison_type, arguments, category_name):
+        category = self._get_category_or_none()
+        if not category:
+            raise ValueError(f'Category ({category_name}) not found. Please add category before adding the case')
+
+        case = self._get_case_or_none(comparison_type, arguments, category.uuid)
+        return case if case else self._add_case(comparison_type, arguments, category.uuid)
+
+    def get_or_create_category(self, category_name, destination_uuid, is_default=False):
+        category = self._get_category_or_none(category_name)
```

Comment:
This is the `get_or_create_category()` method. My thought process is that modifying anything about the category should be done whoever is calling this function because we might not always want to override what is in the category.

Alternatively, it might make sense to make an `update_or_create_category` method

Reply:
I don't propose to update the category if differing information comes in, but rather issue a warning, because this discrepancy likely means that there is a mistake in the input data.